### PR TITLE
Instrumentation

### PR DESCRIPTION
**Pull request by stuartnelson3**
migration of https://github.com/elastic/beats/pull/30385

### CODE REVIEW

**Comment by stuartnelson3** [comment]
When I applied my changes, there were some unexpected changes: https://github.com/elastic/elastic-agent/pull/180/commits/315a06182e301e4fb000bd79caa74fcbc5bec301

not sure why this would be in my PR but not in this repo.

**Comment by stuartnelson3** [comment]
the PR seems to be failing on linting issues that are already present in main

File: internal/pkg/agent/cmd/run.go [hunk 1]
```diff
@@ -44,6 +47,10 @@ const (
 	agentName = "elastic-agent"
 )
 
+func init() {
+	apm.DefaultTracer.Close()
+}
```
Comment (AndersonQ) [inline internal/pkg/agent/cmd/run.go:52]: [Blocker]
Please move it to the tracer init function. having it here generates a side effect when importing this package that is unrelated to the package.

File: internal/pkg/agent/cmd/run.go [hunk 2]
```diff
@@ -135,22 +142,33 @@ func run(streams *cli.IOStreams, override cfgOverrider) error {
...
+	tracer, err := initTracer(agentName, release.Version(), cfg.Settings.MonitoringConfig)
+	if err != nil {
+		return err
+	}
+	if tracer != nil {
+		defer func() {
+			tracer.Flush(nil)
+			tracer.Close()
+		}()
+	}
```
Comment (AndersonQ) [inline internal/pkg/agent/cmd/run.go:154]: [Suggestion]
It'd be good to log if the APM is enabled or not, either here or on the `initTracer` function.

File: internal/pkg/agent/control/server/server.go [hunk 3]
```diff
@@ -294,6 +303,39 @@ func (s *Server) Pprof(ctx context.Context, req *proto.PprofRequest) (*proto.Ppr
...
+		Result: []*proto.MetricsResponse{metrics},
+	}
+
+	// gather metrics buffer data from all other processes
+	specs := s.getSpecInfo("", "")
+	for _, si := range specs {
+		endpoint := monitoring.MonitoringEndpoint(si.spec, runtime.GOOS, si.rk)
+		client := newSocketRequester(si.app, si.rk, endpoint)
+
+		s.logger.Infof("GATHER METRICS FROM %s", endpoint)
```
Comment (AndersonQ) [inline internal/pkg/agent/control/server/server.go:332]: [Nit]
No need to use all capitals.

File: internal/pkg/agent/operation/operator.go [hunk 4]
```diff
@@ -155,13 +157,17 @@ func (o *Operator) Close() error {
...
 	o.statusReporter.Unregister()
 
-	return o.HandleConfig(configrequest.New("", time.Now(), nil))
+	return o.HandleConfig(context.Background(), configrequest.New("", time.Now(), nil))
 }
 
 // HandleConfig handles configuration for a pipeline and performs actions to achieve this configuration.
-func (o *Operator) HandleConfig(cfg configrequest.Request) (err error) {
+func (o *Operator) HandleConfig(ctx context.Context, cfg configrequest.Request) (err error) {
+	span, ctx := apm.StartSpan(ctx, "route", "app.internal")
```
Comment (AndersonQ) [inline internal/pkg/agent/operation/operator.go:165]: Just double checking, is `route` the right name for this span?

File: internal/pkg/core/monitoring/config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	for name, tc := range tcs {
+		t.Run(name, func(t *testing.T) {
+			in, err := config.NewConfigFrom(tc.in)
+			require.NoError(t, err)
+
+			cfg := DefaultConfig()
+			require.NoError(t, in.Unpack(cfg))
+
+			require.NoError(t, err)
```
Comment (AndersonQ) [inline internal/pkg/core/monitoring/config/config_test.go:60]: Haven't you checked it already on line 55?

File: internal/pkg/core/monitoring/config/config_test.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			require.NotNil(t, cfg)
```
Comment (AndersonQ) [inline internal/pkg/core/monitoring/config/config_test.go:61]: [Suggestion]
If it's checking the result of `in.Unpack(cfg)`, it'd be better to have it placed immediately after the call to `in.Unpack(cfg)`, on line 58.

File: internal/pkg/core/monitoring/config/config_test.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+			instCfg := cfg.APM
+			assert.DeepEqual(t, tc.out, instCfg)
```
Comment (AndersonQ) [inline internal/pkg/core/monitoring/config/config_test.go:63]: [Nit]
Why are you declaring a new variable, `assert.DeepEqual(t, tc.out, cfg.APM)` should yield the same result.

**Review by AndersonQ — CHANGES_REQUESTED** [review]

File: internal/pkg/agent/application/pipeline/stream/operator_stream.go [hunk 8]
```diff
@@ -44,8 +48,13 @@ func (b *operatorStream) Specs() map[string]program.Spec {
 	return nil
 }
 
-func (b *operatorStream) Execute(cfg configrequest.Request) error {
-	return b.configHandler.HandleConfig(cfg)
+func (b *operatorStream) Execute(ctx context.Context, cfg configrequest.Request) (err error) {
+	span, ctx := apm.StartSpan(ctx, "route", "app.internal")
```
[the remark this answers is not in the document]
Comment (ph) [reply internal/pkg/agent/application/pipeline/stream/operator_stream.go:52]: I belive this is correct.

File: internal/pkg/agent/cmd/run.go (hunk 1 above)
Comment (ph) [reply internal/pkg/agent/cmd/run.go:52]: @AndersonQ  I think this is debatable, I am not a fan of `init()`,  but in the case of APM you want to make sure you disable it as early in the process as we can across the project and `init` gives some guarantee of that. Metrics/traces and even logging are always problematic in the context of execution.  Especially in beats or agent, where the start-up sequence can get a bit confusing.

File: internal/pkg/agent/application/pipeline/dispatcher/dispatcher.go [hunk 9]
```diff
@@ -74,7 +76,25 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
...
+	// Creating a child context that carries both the ad.ctx cancelation and
+	// the span from ctx.
+	ctx, cancel := context.WithCancel(ad.ctx)
+	// golint disallows us from discarding the cancel func, even
+	// though we know that the resources for ctx will be released
+	// when ad.ctx is canceled.
+	// ctx is no longer needed after calling acker.Commit(ctx), so we can
+	// safely call cancel().
+	defer cancel()
+	ctx = apm.ContextWithSpan(ctx, span)
```
Comment (ph) [inline internal/pkg/agent/application/pipeline/dispatcher/dispatcher.go:96]: Agree, on the comment here, let's add a `nolint` rule here, https://golangci-lint.run/usage/false-positives/

File: internal/pkg/agent/application/pipeline/router/router_test.go [hunk 10]
```diff
@@ -45,12 +46,13 @@ type notifyFunc func(pipeline.RoutingKey, rOp, ...interface{})
...
 	programs := []program.Program{program.Program{Spec: program.Supported[1]}}
+	ctx := context.Background()
 
 	t.Run("create new and destroy unused stream", func(t *testing.T) {
 		recorder := &recorder{}
 		r, err := New(nil, recorder.factory)
 		require.NoError(t, err)
-		r.Route("hello", map[pipeline.RoutingKey][]program.Program{
+		r.Route(ctx, "hello", map[pipeline.RoutingKey][]program.Program{
 			pipeline.DefaultRK: programs,
```
Comment (ph) [inline internal/pkg/agent/application/pipeline/router/router_test.go:56]: OK to add nolint.

File: internal/pkg/agent/application/pipeline/router/router_test.go [hunk 11]
```diff
@@ -62,7 +64,7 @@ func TestRouter(t *testing.T) {
 		recorder.reset()
 
 		nk := "NEW_KEY"
-		r.Route("hello-2", map[pipeline.RoutingKey][]program.Program{
+		r.Route(ctx, "hello-2", map[pipeline.RoutingKey][]program.Program{
 			nk: programs,
```
Comment (ph) [inline internal/pkg/agent/application/pipeline/router/router_test.go:68]: nolint ok here.

File: internal/pkg/agent/application/pipeline/router/router_test.go [hunk 12]
```diff
@@ -80,7 +82,7 @@ func TestRouter(t *testing.T) {
 		recorder := &recorder{}
 		r, err := New(nil, recorder.factory)
 		require.NoError(t, err)
-		r.Route("hello", map[pipeline.RoutingKey][]program.Program{
+		r.Route(ctx, "hello", map[pipeline.RoutingKey][]program.Program{
 			pipeline.DefaultRK: programs,
```
Comment (ph) [inline internal/pkg/agent/application/pipeline/router/router_test.go:86]: Nolint: ok here.

File: internal/pkg/agent/application/pipeline/router/router_test.go [hunk 13]
```diff
@@ -118,7 +120,7 @@ func TestRouter(t *testing.T) {
...
```
Comment (ph) [inline internal/pkg/agent/application/pipeline/router/router_test.go:124]: golint ok here.

File: internal/pkg/agent/control/server/server.go (hunk 3 above)
Comment (ph) [reply internal/pkg/agent/control/server/server.go:332]: Let's do the change.

**Comment by ph** [comment]
@AndersonQ Do you agree with my last statement?

**Comment by AndersonQ** [comment]
> @AndersonQ Do you agree with my last statement?

Yes, I agree. Indeed for me there was only one thing needed to be addressed, but if you're fine letting it on the `init` function, I'm happy to compromise :)

**Comment by simitt** [comment]
Thanks for the review @ph . 
I'd like to ask @AndersonQ to directly push to this branch with the latest suggested changes. Stuart is on PTO and we would really like to get this in before any other conflicts etc are coming up.

File: internal/pkg/agent/cmd/run.go [hunk 14]
```diff
@@ -374,3 +400,64 @@ func tryDelayEnroll(ctx context.Context, logger *logger.Logger, cfg *configurati
 	logger.Info("Successfully performed delayed enrollment of this Elastic Agent.")
 	return loadConfig(override)
 }
+
+func initTracer(agentName, version string, mcfg *monitoringCfg.MonitoringConfig) (*apm.Tracer, error) {
+	apm.DefaultTracer.Close()
```
Comment (ph) [inline internal/pkg/agent/cmd/run.go:405]: @AndersonQ I believe this should stay in the `init()` func, the tracing needs to be disabled as soon as we can and probably before the `initTracer()` method is called.

**Review by ph — CHANGES_REQUESTED** [review]
Change LGTM, 

I've looked at the CI, i think its a red herring.
```
[2022-03-18T12:10:37.565Z] Timeout, server 3.145.3.184 not responding.
```

File: internal/pkg/agent/cmd/run.go (hunk 14 above)
Comment (AndersonQ) [reply internal/pkg/agent/cmd/run.go:405]: orveral you said the PR is good, so I'm a wee unsure now, is this comment a blocker or not?

My main concern here is that the default tracer closing will be "hidden and lost" on the init function. Any future change might miss it as the whole code for initiating the tracer is here, on `initTracer`. Besides I hardly believe we'd send spans to a wrong server by closing the default tracer here instead of at the packages `init` func.
However if you say it's a blocker, I'm happy to change.

Comment (ph) [reply internal/pkg/agent/cmd/run.go:405]: I get the idea, let me approve this and lets try it out.

**Review by AndersonQ — APPROVED** [review]
